R/IonDb.R

```diff
@@ -0,0 +1,224 @@
+#' 
+#' @param cdb for `IonDb`: `CompDb` object used to construct a `IonDb` object 
+#'     (the content of `cdb` is copied to the `IonDb` object).
+#' 
+#' @param ions for `IonDb`: table to be added to the `IonDb` object (if missing
+#'     an empty table is added). `ions` is required to have the columns 
+#'     "compound_id", "ion_adduct", "ion_mz" and "ion_rt". 
+#'
+#' @param ... additional arguments. Currently not used.
+#'
+#' @author Andrea Vicini, Johannes Rainer
+#'
+#' @examples
+#'
+# # I have to improve/complete this
+# We load a small compound test database based on MassBank which is
+# distributed with this package.
+# cdb <- CompDb(system.file("sql/CompDb.MassBank.sql", package = "CompoundDb"))
+# cdb
+# 
+# # The previous database is read only. We can create a new IonDb object from
+# # cdb and add a ions table to it.
+# 
+# ions = data.frame(compound_id = c(1, 1, 2, 3, 6, 35),
+#                   ion_adduct = c("A", "B", "B", "B", "F", "S"),
+#                   ion_mz = 1:6,
+#                   ion_rt = 1:6) # maybe change this with something more realistic or we should include an example ion table in the package?
+# library(DBI)
+# idb <- IonDb(dbConnect(RSQLite::SQLite(), paste0(tempdir(), "/idb_ex.sqlite")),
+#              cdb, ions)
+# 
+# # It is also possible to load a previously created IonDb passing only the 
+# # connection to the database.
+# idb2 <- IonDb(paste0(tempdir(), "/idb_ex.sqlite"))
+# 
+# # We can also add new ions to the IonDb object through the method insertIon
+# 
+# ## List all available ion variables
+# ionVariables(idb)
+# 
+# ## Extract a data.frame with ion variables for all ions
+# ions(idb)
+# 
+# ## List all database tables and their columns
+# tables(idb)
+# 
+# ## Filtering the database
+# ##
+# ## Get all ions with an exact mass between 310 and 320
+# res <- ions(idb, filter = ~ ion_rt > 3 & ion_mz < 5)
+# res
+# 
+# ## Get all ions that have an H14 in their formula.
+# res <- ions(idb, filter = IonAdductFilter("B", "contains"))
+# res
+NULL
+
+#' @importFrom methods new
+#'
+#' @exportClass IonDb
+.IonDb <- setClass("IonDb",
+                   contains = "CompDb")
+
+#' @importFrom methods validObject
+setValidity("IonDb", function(object) {
+    if (!is.null(object@dbcon))
+        .validIonDb(object@dbcon)
+    else TRUE
+})
+
+#' @importFrom DBI dbListTables dbGetQuery
+.validIonDb <- function(x) {
+    txt <- character()
+    # maybe this is already called automatically by the validation of base class?
+    #txt <- c(txt, .validCompDb(x))            
+    if(!("ms_ion" %in% dbListTables(x)))
+        txt <- c(txt, "Required table 'ms_ion' not found not found in database")
+    else {
+       ions <- dbGetQuery(x, "select * from ms_ion limit 3") 
+       res <- .valid_ion(ions, error = FALSE)
+       if (is.character(res))
+           txt <- c(txt, res)
+    }
+    if (length(txt)) txt else TRUE
+}
+
+#' @export
+#'
+#' @rdname IonDb
+IonDb <- function(x, cdb, ions) {
```

Comment:
We might want to change that function later, but for that I need to use it first "out-in-the-wild" - to determine what the most frequent use case would be...